Refactor refguide-check, take 3

ev-br opened the pull request:
#### Reference issue
<!--Example: Closes gh-WXYZ.-->

Continues and closes gh-16391, supersedes and closes gh-19242

#### What does this implement/fix?
<!--Please explain your changes.-->

Continue gh-19242 by @Sheila-nk : refactor the doctesting part of the refguide-check into a standalone tool, use pytest for running and orchestration. 

So this PR has essentially two parts: 
- heavy lifting done by Sheila (see commit attributions) + additional fixes to docstring examples from gh-19242. A part with pytest ignores was suggested by Lucas in https://github.com/scipy/scipy/pull/19242#discussion_r1431636535, also commit-attributed.

The remaining TODOs follow https://github.com/scipy/scipy/pull/19242#issuecomment-1719964303. What's relevant now is

- [ ] finish plumbing the external tool as git submodule
- [ ] activate checking the code in tutorials
- [ ] drop the obsoleted part of `refguide-check`. I'd prefer to do it in a quick follow-up PR.

On the 'scpdt' tool side, we need to

- [ ] get a better name (doctest_sp, doctest_scipy, scipy_doctest, something else?) --- `doctest-scipy` unless
- [ ] get out of my GH --- will propose to move it to the scipy GH org on the ML

The tool itself is in a bit of a flux at the moment, nothing blocking though. Here's the tracker which is being actively worked on: https://github.com/ev-br/scpdt/issues

The pytest plugin is written by @Sheila-nk with inputs from @melissawm and myself. Thank you Sheila and Melissa!

#### Additional information
<!--Any additional information you think is important.-->

I am starting to dislike tagging this all as _doctesting_. The goal is not ---never was and will never be --- to mix doctests into unit testing. The goal is to keep the examples in the docs current. If somebody comes up with a catchy name for this, I'm all ears :-).

ev-br committed:
DOC: doctest tutorial/io.rst

The tweaks are:
- the Octave version is different (somebody forgotten to check in
  regenerated files, in 2013)
- the dict order is different, at least on python 3.8

ev-br committed:
DOC: optimize: fix errors in milp and basinhopping docstrings

ev-br committed:
DOC: sparse: fix the module docstring example

ev-br committed:
MAINT: sparse.linalg: fix docstring examples (warnings, deprecations)

ev-br committed:
DOC: sparse.csgraph: bump lobpcg tolerance to 1e-2

ev-br committed:
MAINT: stats.qmc: add __all__ for the doctester

ev-br committed:
MAINT: stats.sampling: add __all__ list for the doctester

ev-br committed:
ENH: doctest: recurse into tutorial subdirs; fix easy erorrs, skip the rest

Sheila-nk committed:
ENH: doctests: add conftest.py filters for pytest based doctesting

ev-br committed:
MAINT: doctest: add pytest_extra_skips

ev-br committed:
MAINT: doctests: allow sparse.linalg.norm emit UserWarnings

This is `lopbcg` having convergence issues, again.

Sheila-nk committed:
DEV: add test --doctests option to dev.py

Sheila-nk committed:
MAINT: more fine-grained ignores in dev.py test --doctests

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

Sheila-nk committed:
CI: run doctests via plugin on CircleCI

[docs only]

ev-br commented on scipy/optimize/_milp.py:
```diff
@@ -322,7 +322,7 @@ def milp(c, *, integrality=None, bounds=None, constraints=None, options=None):
     >>> A = np.array([[-1, 1], [3, 2], [2, 3]])
     >>> b_u = np.array([1, 12, 12])
-    >>> b_l = np.full_like(b_u, -np.inf)
+    >>> b_l = np.full_like(b_u, -np.inf, dtype=float)
```
@mdhaber git blame points at you, so I'd ask about this change. Without it, running the example from the `milp` docstring generates a warning from numpy; adding a dtype above changes the result from `[1.0, 2.0]` to `[2.0, 2.0]`:

```
In [1]: import numpy as np

In [2]: c = -np.array([0, 1])

In [3]: A = np.array([[-1, 1], [3, 2], [2, 3]])
   ...: >>> b_u = np.array([1, 12, 12])
   ...: >>> b_l = np.full_like(b_u, -np.inf)
/home/br/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/numpy/core/numeric.py:407: RuntimeWarning: invalid value encountered in cast
  multiarray.copyto(res, fill_value, casting='unsafe')

In [4]: from scipy.optimize import LinearConstraint
   ...: >>> constraints = LinearConstraint(A, b_l, b_u)

In [5]: integrality = np.ones_like(c)

In [6]: from scipy.optimize import milp
   ...: >>> res = milp(c=c, constraints=constraints, integrality=integrality)
   ...: >>> res.x
Out[6]: array([1., 2.])
```

ev-br commented on scipy/sparse/__init__.py:
```diff
@@ -230,7 +230,6 @@
 
 >>> A = lil_array((1000, 1000))
 >>> A[0, :100] = rand(100)
->>> A[1, 100:200] = A[0, :100]
```
@dschult had to remove this line because otherwise it fails and compains about 1D slices.

ev-br commented on scipy/stats/sampling.py:
```diff
@@ -66,3 +66,8 @@
+__all__ = ["NumericalInverseHermite", "NumericalInversePolynomial",
+           "TransformedDensityRejection", "SimpleRatioUniforms",
+           "RatioUniforms", "DiscreteAliasUrn", "DiscreteGuideTable",
+           "UNURANError", "FastGeneratorInversion"]
```
@chrisb83 git blame points at you: any reason `scipy.stats.sampling` did not have an `__all__` list, and would you object if it gets one?

chrisb83 replied:
I think I followed `stats/qmc.py`. I don't see a problem with the addition of __all__

ev-br reviewed (commented):
Let me flag several tweaks to the docstrings. The fixes are mainly mechanical, may very well be completely wrong, so would appreciate a sanity check from subject experts.

ev-br commented:
Testing `dev.py` options:

- [ ] `$ python dev.py test --doctests -j 4` seems to give all workers all tests instead of distributing them? Needs checking, can be done in a follow-up (`refguide-check` never supported this)
- [ ]  `$ python dev.py test --doctests -t path/to/file` does not work. Can be done in a follow-up, as `refguide-check` never supported this.

One other bikeshed item is the name of the switch: `dev.py test --doctests` (current state) or `dev.py test --doctests-only`?

With regular pytest `pytest --doctests` means run _both_ unit tests _and_ doctests; Here `dev.py test --doctests` skips regular unit tests entirely. OTOH `--doctests-only` is more typing. Thoughts?

ev-br committed:
CI: doctest: skip scipy.sparse.linalg.norm

[docs only]

tupui commented on scipy/stats/qmc.py:
```diff
@@ -233,3 +233,4 @@
 
 """
 from ._qmc import *  # noqa: F403
+from ._qmc import __all__  # noqa: F403
```
Instead of `*`? Isn't it redundant then? But sure no objections otherwise 😃 Whatever works as long we don't change the imports and I can still do `sp.stats.qmc.Sobol`

lucascolley replied:
what is the error without this change? Pretty sure this isn't the ideal solution.

tupui reviewed (commented):
Really cool 🚀 

> I am starting to dislike tagging this all as doctesting. The goal is not ---never was and will never be --- to mix doctests into unit testing. The goal is to keep the examples in the docs current. If somebody comes up with a catchy name for this, I'm all ears :-).

We have smoke-tests, what about `smoke-doc`? 😅

lucascolley commented on scipy/_lib/_testutils.py:
```diff
@@ -72,7 +72,16 @@ def __call__(self, label="fast", verbose=1, extra_argv=None, doctests=False,
-            raise ValueError("Doctests not supported")
+            pytest_args += [
+                "--doctest-modules",
+                "--ignore=scipy/interpolate/_interpnd_info.py",
+                "--ignore=scipy/_lib/array_api_compat",
+                "--ignore=scipy/_lib/highs",
+                "--ignore=scipy/_lib/unuran",
+                "--ignore=scipy/_lib/_gcutils.py",
+                "--ignore=scipy/_lib/doccer.py",
+                "--ignore=scipy/_lib/_uarray",
```
can we get rid of the errors from the files which aren't submodules / vendored?

ev-br replied:
Probably. Together with https://github.com/ev-br/scpdt/issues/94
I'd defer it to a follow-up though.

ev-br replied:
That said, if you feel like tackling it in this PR Lucas, I won't mind at all :-).

lucascolley replied:
don't tempt me... (no promises but maybe)

ev-br committed:
BUG: smoke-docs: fix dev.py test --doctest -t path/to/file

plumb through the doctest collection `strategy`:

- dev.py test --doctests : uses strategy=api
- dev.py -s submodule : uses strategy=api
- dev.py -t scipy/linalg._basic.py : uses strategy=None

This sorta makes sense: if you want to smoke-test a package or a subpackage, only
look at the public API --- this is not changed.

If you smoke-test an _individual file_, you want functions in that file, so the
vanilla doctest finder gives you all of them. IOW, given a single implementation
file, you have no way to tell which functions from this file end up in the scipy
public API.

A possible UX glitch is that `dev.py -t` may pick up extra docstrings (from things
which remain private in SciPy). It may thus report some extra failures,
compared to `dev.py -s submodule/which/has/the/file` --- if the docstrings of these
private objects are not perfect.

Two options: bend over backwards to invent a way to figure out if an object from
a private _module ends up in the public API, or simply document it and move on.
I suspect the former will be hacky and unreliable, so am opting for the latter.

Yet another way around it is to specify an individual object:

```
$ python dev.py test --doctests -t scipy/linalg/_basic.py::scipy.linalg._basic.det
```

note the need to specify the full dotted path.

ev-br committed:
ENH: docs: add a conftest

$ pytest doc/source/tutorial --doctest-glob=*rst --ignore=doc/source/tutorial/examples

1. Add a conftest to scipy/doc/source/tutorials: otherwise the doctesting
customization is not found --- it lives under scipy/scipy/conftest.py.

2. We also need to entirely skip the contents of doc/source/tutorial/examples:
these files are included into *rst tutorials, doctest them with the rest of
*rst.

ev-br committed:
MAINT: docs: fix small issues in tutorials

ev-br committed:
MAINT: doctests: skip broken tutorials

Some of them are broken (io.rst? sampling_XXX?); others need a tooling update.

This way, the output of

$ pytest doc/source/tutorial/ --doctest-glob=*rst --ignore=doc/source/tutorial/examples

is green.

ev-br committed:
MAINT: doctests: a more detailed ignore/skip/xfail doctests

ev-br committed:
DEV: refguide-check: add dev.py smoke-docs command, remove `test --doctests`

ev-br committed:
DEV: doctests: add dev.py smoke-tutorial command

ev-br committed:
CI: run smoke-docs on CircleCI

[docs only]

ev-br committed:
MAINT: appease the linter

ev-br commented:
Here's an update:

- Following up on the ML discussion and Pamphile's suggestion (thanks @tupui), the dev.py plumbing is through a dedicated command: 
 `$ python dev.py smoke-docs`. The command mirrors the `test` command, and accepts much the same arguments: `smoke-docs -s submodule`, `smoke-docs -t path/to/file::Klass::method` etc.

- To smoke-test the tutorial, there's a separate command `$ python dev.py smoke-tutorials`

- smoke-testing the tutorials now includes subfolders--- those in `doc/source/tutorial/{stats, interpolate}` have never been checked by refguide-check.  This PR now includes several easy fixes and xfails for less obvious errors.

ev-br commented on doc/source/tutorial/stats/sampling_pinv.rst:
```diff
@@ -173,22 +174,22 @@ PDF evaluations increase during setup for small values of ``u_resolution``.
     ...                                  random_state=urng)
     >>> dist.callbacks
-    11454
+    11454       # may vary
```
@tirthasheshpatel git blame point to you in these sampling_*rst files. Could you take a look over my tweaks to see if they mask real issues?

ev-br commented on scipy/conftest.py:
```diff
@@ -236,3 +236,133 @@ def skip_if_array_api(xp, request):
+    "ND_regular_grid.rst": "ReST parser limitation",
+    "ND_unstructured.rst": "matplotlib deprecation",
+    "extrapolation_examples.rst": "ReST parser limitation",
+    "sampling_pinv.rst": "__cinit__ unexpected argument",
```
@tirthasheshpatel had to xfail the whole tutorial, looks like the code has changed after the docs were written?

ev-br reviewed (commented):
Flagging a couple of possible issues in the tutorials.

Fixing those would be an enhancement, since these were never refguide-checked. So fixes may equally well come in this PR (send me a PR or let me know what to cherry-pick) or as follow-ups.

ev-br committed:
DOC: tutorial: fix MPL deprecation warnings

ev-br committed:
CI: split smoke-docs CI run into separate rubrics

[docs only]

ev-br committed:
DOC: io: adapt to NumPy 2.0 scalar repr

While at it, use csr_array not csr_matrix (cosmetics).

ev-br committed:
DOC: sparse.csgraph: adjust docstrings to NumPy 2.0 scalar repr

ev-br committed:
DOC: ndimage: adjust docstring to NumPy 2.0 scalar repr

ev-br committed:
DOC: stats: adjust docstring to NumPy 2.0 scalar repr

Had to remove Ellipsis matching on numeric values in several
places.

[docs only]

ev-br committed:
DOC: tutorial: fix dev.py smoke-tutorial

[docs only]

ev-br committed:
MAINT: stats: fix doctesting infra for NumPy < 2

ev-br committed:
MAINT: refguide-check: add skips to make it run under numpy < 2

Doctestrings were updated for NumPy 2.0, scalar reprs have changed.
This is a minimal fix, just to avoid `dev.py refguide-check` erroring out.

[docs only]

ev-br committed:
lint

ev-br committed:
TST: do not fail if scpdt is not available

ev-br committed:
TST: _lib: smoke-docs may use warnings

ev-br committed:
MAINT: appease mypy for smoke-docs

lucascolley commented:
CI looks happy 👀

ev-br commented:
Thanks Lucas. Now that CI is green, the plugin repository has been moved to https://github.com/scipy/scipy_doctest, as discussed on the mailing list. Could you check that you've a green button on that repository please?

ev-br committed:
DOC: smoke-docs: adjust the plugin repo path/name

tupui commented:
Should this repo also be name smoke-docs?

ev-br commented:
ATM the repository and the package are `scipy_doctest` and the dev.py stanzas are `$ python dev.py smoke-docs -s submodule` and `$ python dev.py smoke-tutorial`. The former supports the usual `-s`, `-t`, `-v`, `-- -x` knobs etc. Good enough?

tupui commented:
Ah ok 👍 yes all good

lucascolley commented:
Very minor but is there a reason that "smoke-docs" is plural while "smoke-tutorial" is not?

lucascolley commented:
> Could you check that you've the green button on that repository please?

Doesn't look like it.![image](https://github.com/scipy/scipy/assets/51488791/2569d7b7-d6d6-4d34-84e4-978e09999ba2)

ev-br commented:
> Doesn't look like it.

Hmmm. That's not right. Somehow the 'settings' tab is missing after transferring the repo, both in the scipy org and in my personal GH account. A workaround from 2020, https://github.com/orgs/community/discussions/22222, does not seem to work for me either, the `/settings` endpoint 404s for me.

Let's give a bit in a hope there a database to sync or something.

OTOH, I now see that the 'settings' tab is missing on some repos in the scipy org for me, but not all. Which suggests there is some action to be taken by somebody with the admin-level access. I'll follow up.

rgommers commented:
Should be fixed now, it was a matter of (1) adding the `maintainers` team and (2) upgrading @ev-br's rights on that repo to `admin` (not sure why that's not the default after transferring a repo). I also enabled branch protection for `main`.

lucascolley commented:
Yep I have permissions now 👍

ev-br commented:
Thank you Ralf for sorting this out!

ev-br committed:
MAINT: smoke-tutorials (plural)

[docs only]

ev-br commented:
> Very minor but is there a reason that "smoke-docs" is plural while "smoke-tutorial" is not?

Good point! `smoke-tutorials` it is, updated.

ev-br commented:
I'm planning to merge this in about a week unless further comments.

tupui reviewed (approved):
Go go go 👍

ev-br committed:
DOC: update to the sparse repr change

ev-br committed:
CI: install the released scipy-doctest

[docs only]
